File: app/src/main/java/dev/pitlor/smssync/adapters/MessagesAdapter.java
package dev.pitlor.smssync.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.pitlor.sms.Message;
import dev.pitlor.smssync.R;
import dev.pitlor.smssync.databinding.MessagesListItemBinding;
import dev.pitlor.smssync.viewmodels.ConversationsListItemViewModel;

public class MessagesAdapter extends RecyclerView.Adapter<MessagesAdapter.MessageViewHolder> {
    private Fragment fragment;
    private List<Message> messages;
    private LayoutInflater layoutInflater;

    public MessagesAdapter(Fragment fragment, List<Message> messages) {
        this.fragment = fragment;
        this.messages = messages;
        layoutInflater = (LayoutInflater) fragment.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @NonNull
    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        MessagesListItemBinding view = DataBindingUtil
                .inflate(layoutInflater, R.layout.messages_list_item, parent, false);
        return new MessageViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder holder, int position) {
        Message message = messages.get(position);
        holder.setMessage(message);
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    class MessageViewHolder extends RecyclerView.ViewHolder {
        private MessagesListItemBinding itemView;

        public MessageViewHolder(@NonNull MessagesListItemBinding itemView) {
            super(itemView.getRoot());
            this.itemView = itemView;
            itemView.setViewModel(new ViewModelProvider(fragment).get(ConversationsListItemViewModel.class));
        }

        public void setMessage(Message message) {
            itemView.getViewModel().setMessage(message);
        }
    }
}

File: app/src/main/java/dev/pitlor/smssync/adapters/MessagesAdapter.kt
package dev.pitlor.smssync.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.RecyclerView
import dev.pitlor.smssync.datasources.Message
import dev.pitlor.smssync.R
import dev.pitlor.smssync.adapters.MessagesAdapter.MessageViewHolder
import dev.pitlor.smssync.databinding.ListItemMessagesBinding
import dev.pitlor.smssync.viewmodels.ConversationsListItemViewModel

class MessagesAdapter(private val fragment: Fragment) : RecyclerView.Adapter<MessageViewHolder>() {
    private val layoutInflater: LayoutInflater = fragment
        .requireContext()
        .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    private var messages = emptyList<Message>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        val view: ListItemMessagesBinding = DataBindingUtil
            .inflate(layoutInflater, R.layout.list_item_messages, parent, false)
        return MessageViewHolder(view)
    }

    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        holder.setMessage(messages[position])
    }

    override fun getItemCount(): Int {
        return messages.size
    }

    fun setMessages(messages: List<Message>) {
        this.messages = messages
        notifyDataSetChanged()
    }

    inner class MessageViewHolder(val binding: ListItemMessagesBinding) : RecyclerView.ViewHolder(binding.root) {
        private val viewModel by fragment.viewModels<ConversationsListItemViewModel>()

        fun setMessage(message: Message) {
            viewModel.message = message
        }

        init {
            binding.viewModel = viewModel
        }
    }

}